railties/lib/rails/railtie.rb

```diff
@@ -191,7 +191,25 @@ def configure(&block)
         instance.configure(&block)
       end
 
+      def <=>(other) # :nodoc:
+        load_index <=> other.load_index
+      end
+
+      def inherited(subclass)
+        subclass.increment_load_index
+        super
+      end
+
+      protected
+        attr_reader :load_index
+
+        def increment_load_index
+          @@load_counter ||= 0
+          @load_index = (@@load_counter += 1)
+        end
```

Comment:
Here I assume that we don't need a mutex because railties should be defined at load time, so we should be protected by the `Kernel#require` mutex.